vector: created VectorKind and Add with vectors now calls VectorAdd

Henrique-S123 opened the pull request:
#### References to other Issues or PRs

Fixes #26121

#### Brief description of what is fixed or changed

Created a VectorKind class, which is the kind for all vector objects in Sympy. The result of operations between vectors, such as Cross, also has VectorKind.
Added a postprocessor for Add with respect to Vector class to make sure that when calling Add with vectors as arguments the result is an instance of VectorAdd.
Added tests to verify all this.

Example:

In [1]: Sys = CoordSys3D('Sys')

In [2]: Sys.i.kind

Out [2]: VectorKind

In [3]: isinstance(Add(Sys.i, Sys.j), VectorAdd)

Out [3]: True

#### Other comments


#### Release Notes

<!-- BEGIN RELEASE NOTES -->

* vector
  * Added VectorKind
  * Add with vectors now calls VectorAdd

<!-- END RELEASE NOTES -->

Henrique-S123 committed:
vector: created VectorKind and added tests

Created the VectorKind class, which is a subclass
of the Kind class. Every vector object in Sympy has
this kind. To ensure that, added a @property in Vector
class that returns VectorKind().

Example

In [1]: Sys = CoordSys3D('Sys')

In [2]: Sys.i.kind

Out [2]: VectorKind

JotacBravo committed:
vector: fixed Add for vectors

Fixed the Add class to work with vectors.
Add of two vectors now invokes the VectorAdd class.

Example

In [1]: Sys = CoordSys3D('Sys')

In [2]: isinstance(Add(Sys.i, Sys.j), VectorAdd)

Out [2]: True

Henrique-S123 committed:
vector: added tests for VectorKind and for Add dispatching to VectorAdd

Added tests to assure that base vectors and operations involving them
have VectorKind. Also, added tests to assure that calls to Add
with vectors as arguments result in instances of VectorAdd.

Henrique-S123 commented:
The code-quality test fails due to an unused import, which I already removed, and due to referring to classes like Cross or VectorMul, which were not imported, in the VectorKind class docstring.

Henrique-S123 committed:
vector: removed undefined names in VectorKind docstring

References to Cross, VectorAdd, VectorMul and VectorZero
were causing undefined names errors when running flake8,
so just removed it.

Henrique-S123 committed:
vector: fixed kind property definition

The previous definition of the kind property was
causing an error with the mypy tool. It is fixed now.

oscarbenjamin commented on sympy/vector/tests/test_vector.py:
```diff
@@ -51,6 +53,44 @@ def test_vector_sympy():
+    assert VectorZero().kind is VectorKind()
+
+    assert v1.projection(v2).kind is VectorKind()
+    assert v2.projection(v1).kind is VectorKind()
+
+def test_vectoradd():
```
There should be two blank lines between top-level functions, classes etc.

Henrique-S123 replied:
One more blank line was added

oscarbenjamin commented:
I wonder if `VectorKind` should be a parametrised kind like `MatrixKind` is i.e. the kind should be parametrised by the kind of the coefficients.

JotacBravo commented:
> I wonder if `VectorKind` should be a parametrised kind like `MatrixKind` is i.e. the kind should be parametrised by the kind of the coefficients.

Initially, we considered parameterizing `VectorKind` similarly to how `MatrixKind` is parameterized by the kind of coefficients. However, given the specific context in which vectors are used within SymPy, it seems that vectors only interact with numerical coefficients, which makes an additional parameter for element type unnecessary. That said, we are open to revisit this decision and implement parameterization if you believe it is better.

oscarbenjamin commented:
I just tried with this PR and I get:
```python
In [1]: from sympy.vector import *

In [2]: C = CoordSys3D('N')

In [3]: C.x.kind
Out[3]: UndefinedKind
```
I suppose that that should be NumberKind if we are assuming that the coefficients are always numbers.

JotacBravo commented:
> I just tried with this PR and I get:
> 
> ```python
> 
> In [1]: from sympy.vector import *
> 
> 
> 
> In [2]: C = CoordSys3D('N')
> 
> 
> 
> In [3]: C.x.kind
> 
> Out[3]: UndefinedKind
> 
> ```
> 
> I suppose that that should be NumberKind if we are assuming that the coefficients are always numbers.

You are right. Thank you for your feedback. My colleague and I will work on resolving this issue.

Henrique-S123 committed:
scalar: kind for BaseScalar is now defined

Kind for BaseScalar was not defined, so it
was UndefinedKind. Now, it is defined as NumberKind.

Henrique-S123 commented:
C.x is an instance of a BaseScalar, for which kind was not defined. Since it is a subclass of Basic, C.x.kind defaults to UndefinedKind.
It is now defined as NumberKind.

oscarbenjamin commented on sympy/vector/vector.py:
```diff
@@ -344,6 +347,24 @@ def _div_helper(one, other):
+        vec_class = {Add: VectorAdd}[cls]
+        vectors = []
+        for term in expr.args:
+            if term.kind is VectorKind():
```
If `VectorKind` is going to be a singleton (i.e. not parametrised) then it would be better to create it as a singleton like for NumberKind:
https://github.com/sympy/sympy/blob/06dbac029e785296ee3bbfb1001266806d94e98b/sympy/core/kind.py#L162

Henrique-S123 replied:
VectorKind is now a parametrised Kind.

oscarbenjamin replied:
Should this not check something like `if isinstance(term.kind, VectorKind)` because the kind might not be `VectorKind(NumberKind)`?

Henrique-S123 replied:
Yes, it should. I'll change it now.

oscarbenjamin commented:
I'm not sure about having `VectorKind` as a non-parametrised kind. It seems like there could be things that are not NumberKind such as non-dimensionless coefficients:
```python
In [4]: from sympy.physics.units import meter

In [5]: from sympy.vector import *

In [6]: R = CoordSys3D('R')

In [7]: v = 2*meter*R.i

In [8]: v
Out[8]: (2⋅meter) i_R

In [9]: (2*meter).kind
Out[9]: UndefinedKind
```
Currently there is no kind for dimensional quantities but I would expect that the kind for `2*meter` would not be `NumberKind`.

I imagine that there are other cases where the coefficients might not be `NumberKind` as well.

Henrique-S123 commented:
You are right. We didn't think of cases like that when we decided to implement VectorKind as a non-parametrised kind. We'll work on making it a parametrised kind as soon as possible. Thank you for your help and guidance.

Henrique-S123 committed:
vector: VectorKind is now a parametrised kind

There were some cases where the element of a vector
could have a kind other than NumberKind.
So, added a parameter element_kind to the VectorKind
class.

Henrique-S123 committed:
vector: updated VectorKind docstring

After making VectorKind a parametrised kind, we forgot
to update the docstring. It is updated now.

Henrique-S123 commented:
Me and my colleague have already changed VectorKind to a parametrised kind. We also updated the tests and VectorKind class docstring.
For some reason, the authors check fails, and we can't understand why. Can you shed some light on why is it failing?

oscarbenjamin commented:
> the authors check fails, and we can't understand why.

You need to merge with master because the AUTHORS file has been updated while this PR was open.

Henrique-S123 committed:
vector: fixed kind check in vector postprocessor
